File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/general/YearAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.general;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

import com.truveta.opentoken.attributes.BaseAttribute;
import com.truveta.opentoken.attributes.validation.RegexValidator;
import com.truveta.opentoken.attributes.validation.SerializableAttributeValidator;

/**
 * Represents a generic year attribute.
 * 
 * This class extends BaseAttribute and provides functionality for working with
 * year fields. It recognizes "Year" as a valid alias for this attribute type.
 * 
 * The attribute performs normalization on input values by trimming whitespace
 * and validates that the year is a 4-digit year format.
 */
public class YearAttribute extends BaseAttribute {

    private static final String NAME = "Year";
    private static final String[] ALIASES = new String[] { NAME };

    /**
     * Regular expression pattern for validating year format.
     * 
     * This regex validates that the input is:
     * - A 4-digit year
     * - Optionally with leading/trailing whitespace
     */
    private static final String YEAR_REGEX = "^\\s*\\d{4}\\s*$";
    private static final Pattern YEAR_PATTERN = Pattern.compile("\\d{4}");

    /**
     * Constructor for YearAttribute with no additional validators.
     */
    public YearAttribute() {
        this(null);
    }

    /**
     * Protected constructor allowing subclasses to add additional validators.
     * 
     * @param additionalValidators list of additional validators to apply, or null
     */
    protected YearAttribute(List<SerializableAttributeValidator> additionalValidators) {
        super(buildValidators(additionalValidators));
    }

    /**
     * Builds the list of validators by combining the base regex validator with any additional validators.
     * 
     * @param additionalValidators optional additional validators
     * @return combined list of validators
     */
    private static List<SerializableAttributeValidator> buildValidators(
            List<SerializableAttributeValidator> additionalValidators) {
        List<SerializableAttributeValidator> validators = new ArrayList<>();
        validators.add(new RegexValidator(YEAR_REGEX));
        if (additionalValidators != null) {
            validators.addAll(additionalValidators);
        }
        return validators;
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String[] getAliases() {
        return ALIASES;
    }

    @Override
    public String normalize(String value) {
        if (value == null) {
            throw new IllegalArgumentException("Year value cannot be null");
        }

        // Trim whitespace and validate it's a number
        String trimmed = value.trim();

        // Check if it matches the 4-digit year format using pre-compiled Pattern
        if (!YEAR_PATTERN.matcher(trimmed).matches()) {
            throw new IllegalArgumentException("Invalid year format: " + value);
        }

        try {
            int year = Integer.parseInt(trimmed);
            // Return the year as a string
            return String.valueOf(year);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Invalid year format: " + value);
        }
    }

}

File: lib/python/opentoken/src/main/opentoken/attributes/general/year_attribute.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from typing import List
import re
from opentoken.attributes.base_attribute import BaseAttribute
from opentoken.attributes.validation import RegexValidator
from opentoken.attributes.validation.serializable_attribute_validator import SerializableAttributeValidator


class YearAttribute(BaseAttribute):
    """Represents a generic year attribute.

    This class extends BaseAttribute and provides functionality for working with
    year fields. It recognizes "Year" as a valid alias for this attribute type.

    The attribute performs normalization on input values by trimming whitespace
    and validates that the year is a 4-digit year format.
    """

    NAME = "Year"
    ALIASES = [NAME]

    # Regular expression pattern for validating 4-digit year format
    VALIDATION_PATTERN = re.compile(r"^\d{4}$")

    def __init__(self, additional_validators: List[SerializableAttributeValidator] = None):
        """
        Initialize the YearAttribute with optional additional validators.

        Args:
            additional_validators: Optional list of additional validators to apply
        """
        validation_rules = [RegexValidator(self.VALIDATION_PATTERN)]
        if additional_validators:
            validation_rules.extend(additional_validators)
        super().__init__(validation_rules)

    def get_name(self) -> str:
        """Get the name of the attribute.

        Returns:
            str: The name "Year"
        """
        return self.NAME

    def get_aliases(self) -> List[str]:
        """Get the aliases for the attribute.

        Returns:
            List[str]: A list containing the aliases for this attribute
        """
        return self.ALIASES.copy()

    def normalize(self, value: str) -> str:
        """Normalize the year value by trimming whitespace.

        Args:
            value: The year string to normalize

        Returns:
            str: The trimmed year value

        Raises:
            ValueError: If the year is not a valid 4-digit year
        """
        trimmed = value.strip()
        
        # Validate it's exactly 4 digits
        if not self.VALIDATION_PATTERN.match(trimmed):
            raise ValueError(f"Invalid year format: {value}")
        
        return trimmed
